### PR TITLE
Code Changes for Issue 1957 - Validate Security Definitions

### PR DESCRIPTION
**Pull request by sarangan12**
**Rel Issue**: https://github.com/Azure/autorest/issues/1957 

**Sample Message**
"A swagger must have security definitions and must adhere to the specific structure."
"file:///C:/workspace/autorest/src/modeler/AutoRest.Swagger.Tests/bin/Debug/netcoreapp1.0/Resource/Swagger/Validation/security-definitions-validations.json#/securityDefinitions"

@dsgouda @vishrutshah @veronicagg @salameer Please review and approve

### CODE REVIEW

File: src/modeler/AutoRest.Swagger/Validation/SecurityDefinitionsStructureValidation.cs [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+        public override string MessageTemplate => Resources.SecurityDefinitionsStructureValidation;
+
+        /// <summary>
+        /// Checks for the presence and existence of the security definiton
+        /// </summary>
+        /// <param name="entity"></param>
+        /// <returns></returns>
+        public override bool IsValid(Dictionary<string, SecurityDefinition> securityDefinitions)        
+        {
+            if (securityDefinitions.Count != 1)
```
Comment (dsgouda) [inline src/modeler/AutoRest.Swagger/Validation/SecurityDefinitionsStructureValidation.cs]: So you're enforcing we have securityDefinitions for every spec, correct?

File: src/modeler/AutoRest.Swagger/Validation/SecurityDefinitionsStructureValidation.cs [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+            {
+                return false;
+            }
+
+            foreach(KeyValuePair<string, SecurityDefinition> sd in securityDefinitions)
```
Comment (dsgouda) [inline src/modeler/AutoRest.Swagger/Validation/SecurityDefinitionsStructureValidation.cs]: Simplify this as:
```
return securityDefinitions.Where(sdPair=>!sdPair.Key.EqualsIgnoreCase("azure_auth")||!IsSecurityDefinitionModelValid(securityDefinition)).Any();
```

File: src/modeler/AutoRest.Swagger/Validation/SecurityDefinitionsStructureValidation.cs [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+                    return false;
+                }
+            }
+
+            return true;
+        }
+
+        private bool IsSecurityDefinitionModelValid(SecurityDefinition securityDefinition)
+        {
+            if(
```
Comment (dsgouda) [inline src/modeler/AutoRest.Swagger/Validation/SecurityDefinitionsStructureValidation.cs]: Just do:
```
return securityDefinition.SecuritySchemeType == SecuritySchemeType.OAuth2 && securityDefinition.AuthorizationUrl.Equals("https://login.microsoftonline.com/common/oauth2/authorize", StringComparison.CurrentCultureIgnoreCase) &&  securityDefinition.Flow == OAuthFlow.Implicit && !String.IsNullOrEmpty(securityDefinition.Description) && IsSecurityDefinitionScopesValid(securityDefinition.Scopes)
```

File: src/modeler/AutoRest.Swagger/Validation/SecurityDefinitionsStructureValidation.cs [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+            {
+                return true;
+            }
+
+            return false;
+        }
+
+        private bool IsSecurityDefinitionScopesValid(Dictionary<string, string> scopes)
+        {
+            if(scopes == null || scopes.Count != 1)
```
Comment (dsgouda) [inline src/modeler/AutoRest.Swagger/Validation/SecurityDefinitionsStructureValidation.cs]: or.... 
``` scopes?.Count!=1 ```

File: src/modeler/AutoRest.Swagger/Validation/SecurityDefinitionsStructureValidation.cs [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+            {
+                return false;
+            }
+
+            foreach(KeyValuePair<string, string> scope in scopes)
+            {
+                if(!scope.Key.Equals("user_impersonation", StringComparison.CurrentCultureIgnoreCase) || String.IsNullOrEmpty(scope.Value))
```
Comment (dsgouda) [inline src/modeler/AutoRest.Swagger/Validation/SecurityDefinitionsStructureValidation.cs]: You could rewrite this exactly like the Linq statement suggested above

File: src/modeler/AutoRest.Swagger/Validation/SecurityDefinitionsStructureValidation.cs (hunk 1 above)
Comment (sarangan12) [inline src/modeler/AutoRest.Swagger/Validation/SecurityDefinitionsStructureValidation.cs]: Yes

File: src/modeler/AutoRest.Swagger/Validation/SecurityDefinitionsStructureValidation.cs (hunk 2 above)
Comment (sarangan12) [inline src/modeler/AutoRest.Swagger/Validation/SecurityDefinitionsStructureValidation.cs]: You are really getting into LINQ.... :-) :-) Changed the code to use LINQ in all the places

File: src/modeler/AutoRest.Swagger/Validation/SecurityDefinitionsStructureValidation.cs [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+            return securityDefinitions.Where(sdPair =>
+                sdPair.Key.Equals("azure_auth", StringComparison.CurrentCultureIgnoreCase)
+            &&  IsSecurityDefinitionModelValid(sdPair.Value)).Any();
+        }
+
+        private bool IsSecurityDefinitionModelValid(SecurityDefinition securityDefinition)
+        {
+            return (
+                securityDefinition.SecuritySchemeType == SecuritySchemeType.OAuth2 &&
+                securityDefinition.AuthorizationUrl.Equals("https://login.microsoftonline.com/common/oauth2/authorize", StringComparison.CurrentCultureIgnoreCase) &&
```
Comment (dsgouda) [inline src/modeler/AutoRest.Swagger/Validation/SecurityDefinitionsStructureValidation.cs]: nit: just curious if this is a hardcoded string

Comment (sarangan12) [inline src/modeler/AutoRest.Swagger/Validation/SecurityDefinitionsStructureValidation.cs]: Yes

**Comment by sarangan12** [comment]
Still resolving the build issue.

**Pull request merged**